traefik/templates/gateway.yaml

```diff
@@ -14,25 +17,42 @@ metadata:
 spec:
   gatewayClassName: {{ default "traefik" .Values.gatewayClass.name }}
   listeners:
-    - name: web
-      port: {{ .Values.ports.web.port }}
-      protocol: HTTP
-      {{- with .Values.gateway.namespacePolicy }}
-      allowedRoutes:
-        namespaces:
-          from: {{ . }}
+    {{- range $name, $config := .Values.gateway.listeners }}
+    - name: {{ $name }}
+      {{ $found := false }}
+      {{- range $portName, $portConfig := $.Values.ports -}}
+        {{- if eq $portConfig.port $config.port -}}
+          {{ $found = true }}
+        {{- end -}}
+      {{- end -}}
+      {{ if not $found }}
+        {{- fail (printf "ERROR: port %0.f is not declared in ports" .port ) }}
+      {{- end -}}
+      port: {{ .port }}
+      {{- if not (has .protocol (list "HTTP" "HTTPS" "TCP" "TLS" "UDP")) }}
+        {{- fail "ERROR: protocol must be set to HTTP, HTTPS, TCP, TLS or UDP" }}
+      {{- end }}
+      protocol: {{ .protocol }}
+      {{- with .hostname }}
+      hostname: {{ . }}
       {{- end }}
-    {{- if .Values.gateway.certificateRefs }}
-    - name: websecure
-      port: {{ $.Values.ports.websecure.port }}
-      protocol: HTTPS
-      {{- with .Values.gateway.namespacePolicy }}
+      {{- with .namespacePolicy }}
       allowedRoutes:
         namespaces:
           from: {{ . }}
       {{- end }}
+      {{ if or .certificateRefs .mode }}
       tls:
+        {{ with .mode }}
+        {{- if not (has . (list "Passthrough" "Terminate")) }}
```

Comment:
Here I'm even more in favor of removing the validation, given the Gateway API CRD already declares this as an enum, thus resulting in an error when the user tries to set an invalid value. Therefore I'm inclined to say that we should leave validation here to Kubernetes + CRD + API, rather than this chart.